### ISSUE TITLE
Planning CI/test for the 0.4.0 release

### ISSUE THREAD

**Issue by wainersm** [confidential-containers/community#69]
**Motivation**

Alongside the previous three releases:

 * We left some technical debits behind (e.g., re-use of Kata's tests on operator CI is a hack)
 * We experienced instability (e.g. SEV jobs failing)
 * We released features/support not covered on CI (e.g. AMD SEV-ES support) 

Also there will be more/new demands for 0.4.0. For example, e2e testing for cloud-api-adaptor (a.k.a peer pods).

**Goal**

My assessment is that we ended up with a good list of work items to tackle but that is spread on several places: issues on several repos, slack discussions, people's brain (or hopefully hand-written on paper), etc. So I'm opening this issue so that we can collectively and openly:

 * Flush ideas
 * Transform on actionable items (i.e. in issues)
 * Prioritize the issues
 * Assign to people
 * Finally put on @ariel-adam 's magic planning board.

**Comment by fitzthum** [comment]
A few initial comments, 

First, using Kata's CI for the operator is really a political/business issue. CNCF won't really support Jenkins and we don't want to use GHA. One path forward is to find someone who wants to donate a node to be a Jenkins server. We would also need to find some resources to run the non-tee tests. This could be on a cloud somewhere or we might consider running these tests in VMs on one of the nodes that is running a baremetal test. I haven't had much luck finding these sorts of resources, but we could probably hunt them down. 

Second, here are the CI changes that come to mind for SEV. 
- Collapse operator and Kata tests into one (sev.bats) -- this already happened. 
- Caching for SEV components -- Unmesh has PR for OVMF. Initrd has dependency on kernel build for module, which is suboptimal.
- Add test for SEV-ES - this test should be more or less trivial but it is waiting behind Jim's annotation PR, which is waiting on some changes to the CI that I am working on.
- Add test for signed images with SEV - This is a little bit more involved, but I think Alex is looking into it.
- Add memory dump test -- Ryan and I have long wanted to add a test that searches the memory for sensitive information. Not a huge priority, but it could come soon. 
- SNP tests. There are a three different tests for the three phases of SNP support. 

I will leave discussion of the spurious network issues to @ryansavino

**Comment by fidencio** [comment]
> and we don't want to use GHA

We actually **cannot** use GHA as it doesn't support nested virtualisation AFAIU.
If it does support, I'm all in for switching to that (not really needed for CC, but also relevant: https://github.com/kata-containers/ci/issues/529)

**Comment by stevenhorsman** [comment]
A comment I'd have on the SEV CI tests is that we already have testing for signature that the other platforms use. I'd much rather see the SEV testing move to run the existing tests and thereby give SEV the same test coverage as everything else and stay in sync with them when updates then add even more duplication, so I'd like to see the move in that direction if at all possible. As it stands I'm not sure how easy it would be to get agreement for the SEV tests (and therefore function) to be merged into `main` as it doesn't seem to follow the same approach as all the other integration tests, so I'd rather see it refactored now than expand and cause a bigger problem later. @ryansavino - we've mentioned this is passing a number of times, do you have any thoughts on the feasibility of this?

**Comment by wainersm** [comment]
> A few initial comments,
>

btw, very good comments, thanks!
 
> First, using Kata's CI for the operator is really a political/business issue. CNCF won't really support Jenkins and we don't want to use GHA. One path forward is to find someone who wants to donate a node to be a Jenkins server. We would also need to find some resources to run the non-tee tests. This could be on a cloud somewhere or we might consider running these tests in VMs on one of the nodes that is running a baremetal test. I haven't had much luck finding these sorts of resources, but we could probably hunt them down.
>

I'm volunteering to co-owner that. For 0.4.0 I will be mostly work with the e2e CI for peer pods and there is an intersection with that because I don't to put yet another job on Kata's Jenkins nor use their sponsored Azure VMs.

I assume it will tracked on  https://github.com/confidential-containers/community/issues/67

> Second, here are the CI changes that come to mind for SEV.
> 
>     * Collapse operator and Kata tests into one (sev.bats) -- this already happened.
>

Cool! Could you please share the issue link?

For the below work items, I added them on the table in the description of this issue. It is missing their issues, owners and priorities (assume 06 and 07 are low; let me know if I am wrong).
 
>     * Caching for SEV components -- Unmesh has PR for OVMF. Initrd has dependency on kernel build for module, which is suboptimal.
> 
>     * Add test for SEV-ES - this test should be more or less trivial but it is waiting behind Jim's annotation PR, which is waiting on some changes to the CI that I am working on.
> 
>     * Add test for signed images with SEV - This is a little bit more involved, but I think Alex is looking into it.
> 
>     * Add memory dump test -- Ryan and I have long wanted to add a test that searches the memory for sensitive information. Not a huge priority, but it could come soon.
> 
>     * SNP tests. There are a three different tests for the three phases of SNP support.
> 
> 
> I will leave discussion of the spurious network issues to @ryansavino

The other day we talked briefly about those errors. Quoting @ryansavino:

```
These are the errors that I've seen that I think are related to AMD network issues:

* openssl connect error
* spurious network error
* k8s apt key install error
* docker repository add error
```

@ryansavino do we already have issues for those errors?

**Comment by fitzthum** [comment]
@stevenhorsman 

I'm not sure it makes sense to use the existing signature tests for SEV. The signature test that we have in mind uses the KBS to get the signature files after an attestation has taken place. Don't the current tests just put the signature information in the image? I guess it wouldn't hurt to also run those tests with SEV, but it wouldn't be testing attestation codepath. It's also not really a suggested use of CoCo with SEV.

Similarly I guess we could also run the existing authenticated registry tests with SEV, but really we should be getting that information via the KBS. Authenticated registries haven't really been on my mind, but we probably should add that test at some point. It should be fairly similar to the signature test.

In general, I think there is probably more code sharing that could be done between the tests, but I am a little bit wary of thinking of the non-tee tests as the definitive tests given that they don't have any attestation code. There are also a few pieces there that I find a little bit weird about the non-tee tests, like continuing to use Jakob's ssh-demo for the encrypted image (and checking out the docs repo to get it). It also does a few things that are redundant when running on the operator CI and directly modifies the kata config and the image (some places this is inevitable, but we are hoping to avoid both with the SEV CI). To be clear, I think non-tee tests are solid. The code looks lovely and they have really led the way for the project, but I get a bit defensive when you say that the SEV tests couldn't be accepted in main.

A few things I think we can do to converge would be to look into using some of the existing library functions, although I think many of them would have to change some, and to use the same images between tests. If there are other specific issues with the code, please let us know.

**Comment by wainersm** [comment]
> > and we don't want to use GHA
> 
> We actually **cannot** use GHA as it doesn't support nested virtualisation AFAIU. If it does support, I'm all in for switching to that (not really needed for CC, but also relevant: [kata-containers/ci#529](https://github.com/kata-containers/ci/issues/529)). The thing we'd do for Kata Containers is to use our Azure sponsorship with GHA, but one thing we cannot be doing is using Kata Containers' Azure sponsorship to run the Confidential Containers CIs).

Thanks for the input @fidencio ! btw, we will be tracking that discussion on https://github.com/confidential-containers/community/issues/67 on CoCo side.

Any other work I think we should include on 0.4?

**Comment by fitzthum** [comment]
@wainersm 

Many of those things have issues already. I will dig them up (was being a bit lazy yesterday) or make them and add to table.

**Comment by fitzthum** [comment]
>  We actually cannot use GHA as it doesn't support nested virtualisation AFAIU.
If it does support, I'm all in for switching to that (not really needed for CC, but also relevant: https://github.com/kata-containers/ci/issues/529). The thing we'd do for Kata Containers is to use our Azure sponsorship with GHA, but one thing we cannot be doing is using Kata Containers' Azure sponsorship to run the Confidential Containers CIs).

I think that it would not be impossible to use GHA. At the very least we could use a self-hosted runner for non-tee tests and do the virtualization ourselves. There are also some rumors about beta support for nested virt, but it's hard to find details. My sense is that this would be more messing around than it is worth, especially if we were trying to do it with so-called CNCF resources.

**Comment by stevenhorsman** [comment]
> @stevenhorsman
> 
> I'm not sure it makes sense to use the existing signature tests for SEV. The signature test that we have in mind uses the KBS to get the signature files after an attestation has taken place. Don't the current tests just put the signature information in the image? I guess it wouldn't hurt to also run those tests with SEV, but it wouldn't be testing attestation codepath. It's also not really a suggested use of CoCo with SEV.
> 
> Similarly I guess we could also run the existing authenticated registry tests with SEV, but really we should be getting that information via the KBS. Authenticated registries haven't really been on my mind, but we probably should add that test at some point. It should be fairly similar to the signature test.

Thanks for the comments and sorry that I wasn't clear. What I'd intended to imply, (and I think the path that TDX & EAA are working on) is to improve those tests to support the remote KBS's as well, so we could have the same test scenarios for non-TEE with offline_fs_kbc, TDX with EAA_kbc and SEV with online_sev_kbc, but the specific data and images probably would differ for some of the cases.

> In general, I think there is probably more code sharing that could be done between the tests, but I am a little bit wary of thinking of the non-tee tests as the definitive tests given that they don't have any attestation code. There are also a few pieces there that I find a little bit weird about the non-tee tests, like continuing to use Jakob's ssh-demo for the encrypted image (and checking out the docs repo to get it). It also does a few things that are redundant when running on the operator CI and directly modifies the kata config and the image (some places this is inevitable, but we are hoping to avoid both with the SEV CI). To be clear, I think non-tee tests are solid. The code looks lovely and they have really led the way for the project, but I get a bit defensive when you say that the SEV tests couldn't be accepted in main.

I in no way mean to imply that the current tests are perfect as they definitely aren't, I'd just prefer for us to converge so we can make everything better with enhancements we make, rather than having duplicated library code and test scenarios in multiple places so improvement to one is not reflected in others.

**Comment by fitzthum** [comment]
@stevenhorsman 

> Thanks for the comments and sorry that I wasn't clear. What I'd intended to imply, (and I think the path that TDX & EAA are working on) is to improve those tests to support the remote KBS's as well, so we could have the same test scenarios for non-TEE with offline_fs_kbc, TDX with EAA_kbc and SEV with online_sev_kbc, but the specific data and images probably would differ for some of the cases.

Yeah it's true that the codepaths for the other online KBS are wildly different. I'm not even 100 percent sure where those tests are. Would be good to bring them together somehow.

> I in no way mean to imply that the current tests are perfect as they definitely aren't, I'd just prefer for us to converge so we can make everything better with enhancements we make, rather than having duplicated library code and test scenarios in multiple places so improvement to one is not reflected in others.

Yeah I think it would be good to have more shared library code. We can take a look at extracting some of the stuff we use for SEV. A lot of the complexity in those tests is SEV specific stuff, but there are definitely a few things that could be shared.

**Comment by ryansavino** [comment]
> savino do we already have issues for those errors?

I'm tracking these internally at AMD as we investigate potential causes. I can provide updates here.

**Comment by stevenhorsman** [comment]
> > A comment I'd have on the SEV CI tests is that we already have testing for signature that the other platforms use. I'd much rather see the SEV testing move to run the existing tests and thereby give SEV the same test coverage as everything else and stay in sync with them when updates then add even more duplication, so I'd like to see the move in that direction if at all possible. As it stands I'm not sure how easy it would be to get agreement for the SEV tests (and therefore function) to be merged into `main` as it doesn't seem to follow the same approach as all the other integration tests, so I'd rather see it refactored now than expand and cause a bigger problem later. @ryansavino - we've mentioned this is passing a number of times, do you have any thoughts on the feasibility of this?
> 
> @stevenhorsman are you referring to the k8s tests? [5358](https://github.com/kata-containers/tests/pull/5358) removed the custom SEV run.sh and converged all testing to the sev.bats file. Also, the target now follows the same pattern as the other kubernetes targets in the Makefile. This means that all the other K8S_TEST_UNION tests are now running for the SEV tests as well. Also, no more custom installing k8s and cleanup in our scripts. So I think this is done. Am I missing some other conformity to the integration tests that we need to address?

What I'm referring to a couple of things:
1. the SEV tests running the same test cases as the other platforms (with the required customisations of course, similar to the approach that EAA is taking e.g.: https://github.com/kata-containers/tests/blob/b76b3fa4dfc08e72bf5e30552393d125f061649b/integration/confidential/lib.sh#L295-L306), so that we can share them across platforms. As stated above, I completely agree that the current tests aren't perfect, but that's one of the main motivations to get them consolidated so that improvements to a single platform helps all of them. 
2. Removing all the duplicated code in `sev.bats` e.g. at an initial glance I can see methods: `get_version`, `configure_containerd`, `kubernetes_wait_for_pod_ready_state`, `get_container_id` etc. From my understanding these are no specific to SEV and I know that we have code for most of them elsewhere. Obviously if we did 1. then it should do this automatically in the refactor process, but I'd be very happy to see this part done first and separately to start the process.

I hope that helps, but let me know if you want to chat through it further.

**Comment by ryansavino** [comment]
@stevenhorsman - Most of the items you mentioned are organizational, related to the consolidation/sharing of methods for setting things up in the scripts. The link you provided points to the methods for setting up signature verification.

I agree, we can organize/consolidate/share the scripting methods better in the sev.bats file and lib shell scripts. We'll start a PR for this. However, you mentioned that the SEV testing is not running all the existing tests and does not have the same test coverage. Are you referring to the signature verification discussed above? Or something else? I want to make sure that the SEV flows are not missing any tests.

@UnmeshDeodhar

**Comment by stevenhorsman** [comment]
However, you mentioned that the SEV testing is not running all the existing tests and does not have the same test coverage. Are you referring to the signature verification discussed above? Or something else? I want to make sure that the SEV flows are not missing any tests.

Maybe I'm out of date, but we have 13 test cases for the other platform's:
```
22:10:51 ok 1 [cc][agent][cri][containerd] Test can launch pod with measured boot enabled
22:11:06 ok 2 [cc][agent][cri][containerd] Test cannot launch pod with measured boot enabled and rootfs modified
22:11:21 ok 3 [cc][agent][cri][containerd] Test can pull an unencrypted image inside the guest without signature config
22:11:36 ok 4 [cc][agent][cri][containerd] Test can pull a unencrypted signed image from a protected registry
22:11:51 ok 5 [cc][agent][cri][containerd] Test cannot pull an unencrypted unsigned image from a protected registry
22:12:06 ok 6 [cc][agent][cri][containerd] Test can pull an unencrypted unsigned image from an unprotected registry
22:12:20 ok 7 [cc][agent][cri][containerd] Test unencrypted signed image with unknown signature is rejected
22:12:36 ok 8 [cc][agent][cri][containerd] Test unencrypted image signed with cosign
22:12:51 ok 9 [cc][agent][cri][containerd] Test unencrypted image with unknown cosign signature
22:13:06 ok 10 [cc][agent][cri][containerd] Test pull an unencrypted unsigned image from an authenticated registry with correct credentials
22:13:20 ok 11 [cc][agent][cri][containerd] Test cannot pull an image from an authenticated registry with incorrect credentials
22:13:34 ok 12 [cc][agent][cri][containerd] Test cannot pull an image from an authenticated registry without credentials
22:13:55 ok 13 [cc][agent][cri][containerd] Test agent API endpoint can be restricted
```
whereas SEV is not running any of them, but has 4 of it's own tests and they are only cover 2/3 of the above cases?:
```
17:52:38 ok 1 [cc][kubernetes][containerd][sev] Test SEV unencrypted container launch success
17:53:03 ok 2 [cc][kubernetes][containerd][sev] Test SEV encrypted container launch failure with INVALID measurement
17:53:11 ok 3 [cc][kubernetes][containerd][sev] Test SEV encrypted container launch success with NO measurement
17:53:21 ok 4 [cc][kubernetes][containerd][sev] Test SEV encrypted container launch success with VALID measurement
```

**Comment by ryansavino** [comment]
Okay got it. Thanks. We'll work to enable these for the SEV flow.

**Comment by wainersm** [comment]
all, I feed the planning board (https://github.com/orgs/confidential-containers/projects/6/views/12) with items 02-05 from the table above. Feel free to add/remove/adapt items on the board.

I also added an item that I will be most of my time working on: https://github.com/confidential-containers/cloud-api-adaptor/issues/599

**Comment by fitzthum** [comment]
Ok, it seems like there is still some confusion about what type of consolidation we want. @UnmeshDeodhar created https://github.com/kata-containers/tests/pull/5449 to add some code sharing between sev.bats and the other tests, but it seems like what @fidencio had hoped for was to use just one set of tests (`agent_image.bats` and `agent_image_encrypted.bats`) and to add SEV cases to the tests that are already there (see 
https://github.com/kata-containers/tests/pull/5449#pullrequestreview-1307828950). Is this correct? 

Obviously we don't just want to run the non-tee tests for SEV because they don't have attestation, but most of the non-tee tests do have SEV counterparts, some of which are already in `sev.bats`. I don't think that we will support measured boot, but otherwise we could create an SEV case for all the non-tee tests. I see the appeal of having some sort of common flow. I think it could really help us understand the flows across different platforms, but keep in mind that a considerable amount of effort has gone into developing `sev.bats` which is finally starting to settle. There are also a number of pending development efforts to support signatures, authenticated registries, etc. Consolidating these tests is going to take time and will probably delay the introduction of new tests as well as reset the clock on the maturity of the existing ones. Is it worth the effort to end up with something that will be functionally the same (but with a ton of switch statements)?

If we are set on this consolidation, should we try to do that sooner (and delay adding new tests) or later (and have more tests to move over). Also keep in mind that we will soon start to have SNP tests, which will be largely disjoint from the SEV tests. The plan had been to create `snp.bats` and make a new SNP job. Should we do something different?

@stevenhorsman

**Comment by UnmeshDeodhar** [comment]
Good point @fitzthum. I was also under the impression that, since SEV tests require additional steps to be done, it will make more sense to keep them separate from the common flow. Otherwise, in the common test, we need multiple switch cases to perform extra steps needed for SEV.

**Comment by fitzthum** [comment]
@UnmeshDeodhar 

> Good point @fitzthum. I was also under the impression that, since SEV tests require additional steps to be done, it will make more sense to keep them separate from the common flow. Otherwise, in the common test, we need multiple switch cases to perform extra steps needed for SEV.

I think the big drawback is that right now nobody who isn't working on SEV looks `sev.bats`. Maybe consolidating would help get some more eyes on the code. Realistically, though, people would probably still mainly look at the parts that they care about. I have a suspicion that if we did a bunch of work to consolidate them we would end up with a ton of switch statements and then someone down the line would suggest just separating out the SEV tests to make it simpler. I think the most important thing is that we have parity, and that is what we are aiming for either way.

**Comment by fidencio** [comment]
I'd like to have a call, maybe at some point **next week** to have a final decision and agreement on where we're going to, as I have the feeling the discussion here will just be going back and forth on what the Kata Containers maintainers who are working on CoCo prefer, and what the AMD folks want to do.

I'd like to also have AliBaba into the call, as they're adding support to TDX with dragonball, then we discuss what's the best way to move forward, as a community, for the whole set.

For now, let's just put this on hold.

**Comment by fidencio** [comment]
> Ok, it seems like there is still some confusion about what type of consolidation we want. @UnmeshDeodhar created [kata-containers/tests#5449](https://github.com/kata-containers/tests/pull/5449) to add some code sharing between sev.bats and the other tests, but it seems like what @fidencio had hoped for was to use just one set of tests (`agent_image.bats` and `agent_image_encrypted.bats`) and to add SEV cases to the tests that are already there (see
> [kata-containers/tests#5449 (review)](https://github.com/kata-containers/tests/pull/5449#pullrequestreview-1307828950)). Is this correct?

No, this is not correct, and I'd prefer keeping this discussion in the specific PR.  But as we're here ...

The PR consilidates the methods used into `sev.bats` but ignores that there are ways to achieve that using the infrasctruture already provided (as in other methods already existent) as part of the tests.

If there's a way to improve the current methods (and hey, I do think there is), then the PR should cover the improvement on all tests, not only in the SEV ones.

Apart from that, which was the comment that I made as part of my review (and I still don´t get why it was not answered there), there's the point that @stevenhorsman raised and @ryansavino said they'd work on this (see: https://github.com/confidential-containers/community/issues/69#issuecomment-1409026891).  The "
Okay got it. Thanks. We'll work to enable these for the SEV flow." gave me the idea that the consolidation would come enabling the tests as part of the SEV flow.

@fitzthum, is there anything that was missing from my review / take on this that you'd like to have clarified?

**Comment by fitzthum** [comment]
Ok, let's setup a call for next week and really get on the same page. @fidencio I guess you suggestion was not as extreme as I had first thought. You want to not merely copy the code from `sev.bats` to `lib.sh`, but to actually consolidate functionality. Very understandable. And you are saying that rather than join all the bats files together we should move them from `integration/kubernetes/confidential` to `integration/kubernetes` when we are ready to merge into main. Is that right?

That sounds totally compatible with the development that we have planned to bring full coverage to the sev tests.

**Comment by fidencio** [comment]
There are two points, Tobin.

The first one is, and I just discussed this with @UnmeshDeodhar, is that we should create a `kubernetes/libs.sh` and actually consolidate there the common functions (or common code that can be replaced by functions) that are used by the `kubernetes/*.bats`, `kubernetes/confidential/*.bats`, and `kubernetes/sev/*.bats`.  This is what I consider the very first stage of consolidation.   Examples of this are common methods to install / configure containerd, or functions / ways to change the Kata Containers configuration files.  Those things are already there and we shouldn't be doing those in different ways.

Whether we'll get to the point to merge `sevs.bats` into ` agent_image.bats`  is the second point and it's a conversation we'll have to have and decide this as a community.  If the maintainers of the project see the benefit of doing that, that's the way we should go, and I had understood that @ryansavino was okay with this approach based on https://github.com/confidential-containers/community/issues/69#issuecomment-1409026891, but let's have this discussion next week.

One problem that I'm having here is the **perception**, **my perception**, that any review provided by the community is not exactly welcome as it'll  generate more work (and that's generally the case, reviews will end up generating more work, even if it's to simplify the work ...).  We're not here to provide just a rubber-stamp on things that have a green CI, mainly if we foresee those things will generate us headache in the future, and I'd appreciate if everyone could take the "criticism" in the best way possible.

**Comment by fitzthum** [comment]
Ok. Point 1 is very reasonable and I agree that we will have to figure out what is optimal on the second front. 

There's no hesitation to do work on the CI. It's guaranteed that there is going to be a lot of work to get anywhere. There are some real tradeoffs that we will have to think about, but I don't see this as too dramatic. We're all aiming for the best possible solution.

**Comment by surajssd** [comment]
I think we are 4 releases out from 0.4.0 should we close this one?
